utils.go

```diff
@@ -82,31 +83,92 @@ func GetRootlessRuntimeDir(rootlessUID int) (string, error) {
+	return env.tmpPerUserDir
+}
+func (rootlessRuntimeDirEnvironmentImplementation) homeDirGetRuntimeDir() (string, error) {
+	return homedir.GetRuntimeDir()
+}
+func (rootlessRuntimeDirEnvironmentImplementation) systemLstat(path string) (*system.StatT, error) {
+	return system.Lstat(path)
+}
+func (rootlessRuntimeDirEnvironmentImplementation) homedirGet() string {
+	return homedir.Get()
+}
+
+func isRootlessRuntimeDirOwner(dir string, env rootlessRuntimeDirEnvironment) bool {
+	st, err := env.systemLstat(dir)
+	return err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000
+}
+
+// getRootlessRuntimeDirIsolated is an internal implementation detail of getRootlessRuntimeDir to allow testing.
+// Everyone but the tests this is intended for should only call getRootlessRuntimeDir, never this function.
+func getRootlessRuntimeDirIsolated(env rootlessRuntimeDirEnvironment) (string, error) {
+	runtimeDir, err := env.homeDirGetRuntimeDir()
 	if err == nil {
 		return runtimeDir, nil
 	}
-	tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
-	st, err := system.Stat(tmpDir)
-	if err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000 {
-		return tmpDir, nil
+
+	initCommand, err := ioutil.ReadFile(env.getProcCommandFile())
+	if err != nil || string(initCommand) == "systemd" {
+		runUserDir := env.getRunUserDir()
+		if isRootlessRuntimeDirOwner(runUserDir, env) {
+			return runUserDir, nil
+		}
 	}
-	tmpDir = fmt.Sprintf("%s/%d", os.TempDir(), rootlessUID)
-	if err := os.MkdirAll(tmpDir, 0700); err != nil {
-		logrus.Errorf("failed to create %s: %v", tmpDir, err)
-	} else {
-		return tmpDir, nil
+
+	tmpPerUserDir := env.getTmpPerUserDir()
+	if _, err := env.systemLstat(tmpPerUserDir); os.IsNotExist(err) {
+		if err := os.Mkdir(tmpPerUserDir, 0700); err != nil {
+			logrus.Errorf("failed to create temp directory for user: %v", err)
+		} else {
+			return tmpPerUserDir, nil
+		}
+	} else if isRootlessRuntimeDirOwner(tmpPerUserDir, env) {
+		return tmpPerUserDir, nil
 	}
-	home := homedir.Get()
-	if home == "" {
-		return "", errors.Wrapf(err, "neither XDG_RUNTIME_DIR nor HOME was set non-empty")
+
+	homeDir := env.homedirGet()
+	if homeDir == "" {
+		return "", errors.New("neither XDG_RUNTIME_DIR not temp dir nor HOME was set non-empty")
 	}
-	resolvedHome, err := filepath.EvalSymlinks(home)
+	resolvedHomeDir, err := filepath.EvalSymlinks(homeDir)
 	if err != nil {
-		return "", errors.Wrapf(err, "cannot resolve %s", home)
+		return "", errors.Wrapf(err, "cannot resolve %s", homeDir)
 	}
-	return filepath.Join(resolvedHome, "rundir"), nil
+	return filepath.Join(resolvedHomeDir, "rundir"), nil
+}
+
+func getRootlessRuntimeDir(rootlessUID int) (string, error) {
+	return getRootlessRuntimeDirIsolated(
+		rootlessRuntimeDirEnvironmentImplementation{
+			"/proc/1/comm",
+			fmt.Sprintf("/run/user/%d", rootlessUID),
+			fmt.Sprintf("%s/containers-user-%d", os.TempDir(), rootlessUID),
```

Comment:
Note to self: the old path was `"%s/%d"`. I don’t think it matters, this one is at least a bit more attributable.